application.go

```diff
@@ -240,42 +263,59 @@ func (a *Application) Run() error {
-	a.window.SetRefreshCallback(m.glfwRefreshCallback)
-	a.window.SetPosCallback(m.glfwPosCallback)
+	// Setup a new windowManager to handle windows pixel ratio's and pointer
+	// devices.
+	windowManager := newWindowManager(a.config.forcePixelRatio)
+	// force first refresh
+	windowManager.glfwRefreshCallback(a.window)
+	// Attach glfw window callbacks for refresh and position changes
+	a.window.SetRefreshCallback(windowManager.glfwRefreshCallback)
+	a.window.SetPosCallback(windowManager.glfwPosCallback)
 
-	// flutter's PlatformMessage handler is registered through the dart:ui.Window
-	// interface. ui.Window must have at least paint one frame, before any
-	// platfrom message can be corectly handled by ui.Window.onPlatformMessage.
-	glfw.WaitEvents()
+	// TODO: Can this only be done here? Why not in the plugin/glfwPlugin init loop above?
+	for _, p := range a.config.plugins {
+		// Extra init call for plugins that satisfy the PluginTexture interface.
+		if texturePlugin, ok := p.(PluginTexture); ok {
+			err = texturePlugin.InitPluginTexture(texturer)
+			if err != nil {
+				return errors.Wrap(err, "failed to initialize texture plugin"+fmt.Sprintf("%T", p))
+			}
+		}
+	}
 
+	// Attach glfw window callbacks for text input
 	a.window.SetKeyCallback(
 		func(window *glfw.Window, key glfw.Key, scancode int, action glfw.Action, mods glfw.ModifierKey) {
 			defaultTextinputPlugin.glfwKeyCallback(window, key, scancode, action, mods)
 			defaultKeyeventsPlugin.sendKeyEvent(window, key, scancode, action, mods)
 		})
 	a.window.SetCharCallback(defaultTextinputPlugin.glfwCharCallback)
 
+	// Attach glfw window callback for iconification
 	a.window.SetIconifyCallback(defaultLifecyclePlugin.glfwIconifyCallback)
 
-	a.window.SetCursorEnterCallback(m.glfwCursorEnterCallback)
-	a.window.SetCursorPosCallback(m.glfwCursorPosCallback)
-	a.window.SetMouseButtonCallback(m.glfwMouseButtonCallback)
-	a.window.SetScrollCallback(m.glfwScrollCallback)
+	// Attach glfw window callbacks for mouse input
+	a.window.SetCursorEnterCallback(windowManager.glfwCursorEnterCallback)
+	a.window.SetCursorPosCallback(windowManager.glfwCursorPosCallback)
+	a.window.SetMouseButtonCallback(windowManager.glfwMouseButtonCallback)
+	a.window.SetScrollCallback(windowManager.glfwScrollCallback)
+
+	// Shutdown the engine if we return from this function (on purpose or panic)
 	defer a.engine.Shutdown()
 
+	// Handle events until the window indicates we should stop. An event may tell the window to stop, in which case
+	// we'll exit on next iteration.
 	for !a.window.ShouldClose() {
-		glfw.WaitEventsTimeout(0.016) // timeout to get 60fps-ish iterations
-		embedder.FlutterEngineFlushPendingTasksNow()
+		eventLoop.WaitForEvents(func(duration float64) {
+			glfw.WaitEventsTimeout(duration)
+		})
 		defaultPlatformPlugin.glfwTasker.ExecuteTasks()
 		messenger.engineTasker.ExecuteTasks()
 	}
 
+	// TODO: What if the window indicates to stop, but there are tasks left on
```

Comment:
[` Running the task before that time is undefined behavior.`](https://github.com/flutter/engine/blob/6a5cbbdb4fc111b7be711e2d8bfbc8a336895c38/shell/platform/embedder/embedder.h#L1099-L1100)

Should we call `a.engine.Shutdown()` and then, process all of the event in the event loop that are `task event time < current time` ?

Reply:
> Should we call a.engine.Shutdown() and then, process all of the event in the event loop that are task event time < current time ?

Since the engine is stopped, we cannot process the `task event time < current time`.

Reply:
The answer to the question:
> 	// TODO: What if the window indicates to stop, but there are tasks left on
	// the queue?

Is discard them, the engine will shutdown, the loop above already processed the  `task event time < current time`.